internal/syscallfs/rootfs.go

```diff
@@ -0,0 +1,372 @@
+package syscallfs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"strings"
+	"syscall"
+	"time"
+)
+
+func NewRootFS(fs ...FS) (FS, error) {
+	switch len(fs) {
+	case 0:
+		return EmptyFS, nil
+	case 1:
+		if fs[0].GuestDir() == "/" {
+			return fs[0], nil
+		}
+	}
+
+	// Last is the highest precedence, so we iterate backwards to keep runtime
+	// code simpler.
+	ret := &CompositeFS{
+		prefixes:     make([]string, len(fs)),
+		rootPrefixes: map[string]int{},
+		fs:           make([]FS, len(fs)),
+		rootIndex:    -1,
+	}
+	j := 0
+	for i := len(fs) - 1; i >= 0; i-- {
+		// Clean the prefix in the same way path matches will.
+		path := fs[i].GuestDir()
+		pathI, pathLen := stripPrefixesAndTrailingSlash(path)
+		prefix := path[pathI:pathLen]
+		if prefix == "" {
+			if ret.rootIndex != -1 {
+				return nil, errors.New("multiple root filesystems are invalid")
+			}
+			ret.rootIndex = j
+		} else if strings.HasPrefix(prefix, "..") {
+			// ../ mounts are special cased and aren't returned in a directory
+			// listing, so we can ignore them for now.
+		} else if strings.Contains(prefix, "/") {
+			return nil, fmt.Errorf("unsupported GuestDir %s, only root level allowed", path)
+		} else {
+			ret.rootPrefixes[prefix] = j
+		}
+		ret.prefixes[j] = prefix
+		ret.fs[j] = fs[i]
+		j++
+	}
+
+	// Ensure there is always a root match to keep runtime logic simpler.
+	if ret.rootIndex == -1 {
+		return nil, errors.New("you must supply a root filesystem")
```

Comment:
doing this keeps the code heft from being crazy. later changes are easier than doing all possible combinations in one PR. e.g. this would be the case if you never mounted "." (same as "/")